MediaGallerySynchronization/Model/Synchronize.php

```diff
@@ -51,125 +30,56 @@ class Synchronize implements SynchronizeInterface
     private $synchronizerPool;
 
     /**
-     * @var GetAssetsIterator
+     * @var AssetsBatchGenerator
      */
-    private $getAssetsIterator;
+    private $batchGenerator;
 
     /**
-     * @var ResolveNonExistedAssets
+     * @var int
      */
-    private $resolveNonExistedAssets;
-
+    private $batchSize;
+    
     /**
-     * @param IsPathBlacklistedInterface $isPathBlacklisted
-     * @param Filesystem $filesystem
      * @param LoggerInterface $log
      * @param SynchronizerPool $synchronizerPool
-     * @param GetAssetsIterator $assetsIterator
-     * @param ResolveNonExistedAssets $resolveNonExistedAssets
+     * @param AssetsBatchGenerator $batchGenerator
+     * @param int $batchSize
      */
     public function __construct(
-        IsPathBlacklistedInterface $isPathBlacklisted,
-        Filesystem $filesystem,
         LoggerInterface $log,
         SynchronizerPool $synchronizerPool,
-        GetAssetsIterator $assetsIterator,
-        ResolveNonExistedAssets $resolveNonExistedAssets
+        AssetsBatchGenerator $batchGenerator,
+        int $batchSize
     ) {
-        $this->isPathBlacklisted = $isPathBlacklisted;
-        $this->filesystem = $filesystem;
         $this->log = $log;
         $this->synchronizerPool = $synchronizerPool;
-        $this->getAssetsIterator = $assetsIterator;
-        $this->resolveNonExistedAssets = $resolveNonExistedAssets;
+        $this->batchGenerator = $batchGenerator;
+        $this->batchSize = $batchSize;
     }
 
     /**
      * @inheritdoc
      */
     public function execute(): void
     {
-        $failedItems = [];
-        $assetsPaths = [];
-
-        /** @var \SplFileInfo $item */
-        foreach ($this->getAssetsIterator->execute($this->getMediaDirectory()->getAbsolutePath()) as $item) {
-            $path = $item->getPath() . '/' . $item->getFilename();
-            if (!$this->isApplicable($path)) {
-                continue;
-            }
-
-            $assetsPaths[] = $this->getRelativePath($path, $item->getFilename());
-
-            foreach ($this->synchronizerPool->get() as $synchronizer) {
-                if ($synchronizer instanceof SynchronizeFilesInterface) {
+        foreach ($this->synchronizerPool->get() as $synchronizer) {
+            if ($synchronizer instanceof SynchronizeFilesInterface) {
```

Comment:
Let's collect incorrect synchronizers in a variable and throw an exception after the processing if any of them are not implementing the SynchronizeFilesInterface

Reply:
ok